qiskit/primitives/base_sampler.py

```diff
@@ -0,0 +1,179 @@
+  to be bound to the parameters of the quantum circuits.
+  (list of list of float)
+
+The output is the probabilities or quasi-probabilities of bitstrings.
+
+The sampler object is expected to be closed after use or
+accessed within "with" context
+and the objects are called with parameter values and run options
+(e.g., ``shots`` or number of shots).
+
+Here is an example of how sampler is used.
+
+.. code-block:: python
+
+    from qiskit import QuantumCircuit
+    from qiskit.circuit.library import RealAmplitudes
+
+    bell = QuantumCircuit(2)
+    bell.h(0)
+    bell.cx(0, 1)
+    bell.measure_all()
+
+    # executes a Bell circuit
+    with Sampler(circuits=[bell], parameters=[[]]) as sampler:
+        result = sampler(parameters=[[]], circuits=[0])
+        print([q.binary_probabilities() for q in result.quasi_dists])
+
+    # executes three Bell circuits
+    with Sampler([bell]*3, [[]]) as sampler:
+        result = sampler([0, 1, 2], [[]]*3)
+        print([q.binary_probabilities() for q in result.quasi_dists])
+
+    # parametrized circuit
+    pqc = RealAmplitudes(num_qubits=2, reps=2)
+    pqc.measure_all()
+    pqc2 = RealAmplitudes(num_qubits=2, reps=3)
+    pqc2.measure_all()
+
+    theta1 = [0, 1, 1, 2, 3, 5]
+    theta2 = [1, 2, 3, 4, 5, 6]
+    theta3 = [0, 1, 2, 3, 4, 5, 6, 7]
+
+    with Sampler(circuits=[pqc, pqc2], parameters=[pqc.parameters, pqc2.parameters]) as sampler:
+        result = sampler([0, 0, 1], [theta1, theta2, theta3])
+
+        # result of pqc(theta1)
+        print([q.binary_probabilities() for q in result[0].quasi_dists])
+
+        # result of pqc(theta2)
+        print([q.binary_probabilities() for q in result[1].quasi_dists])
+
+        # result of pqc2(theta3)
+        print([q.binary_probabilities() for q in result[2].quasi_dists])
+
+"""
+from __future__ import annotations
+
+from abc import ABC, abstractmethod
+from typing import Callable, Iterable, Sequence
+
+from qiskit.circuit import Parameter, QuantumCircuit
+from qiskit.circuit.parametertable import ParameterView
+from qiskit.exceptions import QiskitError
+
+from .sampler_result import SamplerResult
+
+
+class BaseSampler(ABC):
+    """Sampler base class
+
+    Base class of Sampler that calculates quasi-probabilities of bitstrings from quantum circuits.
+    """
+
+    def __init__(
+        self,
+        circuits: Iterable[QuantumCircuit],
+        parameters: Iterable[Iterable[Parameter]] | None = None,
+    ):
+        """
+        Args:
+            circuits (list[QuantumCircuit]): quantum circuits to be executed
+            parameters (list[list[Parameter]]): parameters of quantum circuits
+                Defaults to `[circ.parameters for circ in circuits]`
+
+        Raises:
+            QiskitError: for mismatch of circuits and parameters list.
+        """
+        self._circuits = tuple(circuits)
+        if parameters is None:
+            self._parameters = tuple(circ.parameters for circ in circuits)
```

Comment:
This iterates through the `circuits` `Iterable` twice, which will fail if the input is a consumable iterable (like a `yield` generator).  The subsequent uses should use `self._circuits`.